storage/src/main/java/tech/pegasys/artemis/storage/V2MapDatabase.java

```diff
@@ -0,0 +1,291 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.storage;
+
+import static tech.pegasys.artemis.datastructures.util.BeaconStateUtil.compute_start_slot_at_epoch;
+import static tech.pegasys.artemis.util.alogger.ALogger.STDOUT;
+
+import com.google.common.primitives.UnsignedLong;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.Collections;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ConcurrentNavigableMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes32;
+import org.mapdb.Atomic;
+import org.mapdb.Atomic.Var;
+import org.mapdb.DB;
+import org.mapdb.DBMaker;
+import org.mapdb.DBMaker.Maker;
+import tech.pegasys.artemis.datastructures.blocks.BeaconBlock;
+import tech.pegasys.artemis.datastructures.state.BeaconState;
+import tech.pegasys.artemis.datastructures.state.Checkpoint;
+import tech.pegasys.artemis.storage.Store.Transaction;
+import tech.pegasys.artemis.storage.utils.Bytes32Serializer;
+import tech.pegasys.artemis.storage.utils.MapDBSerializer;
+import tech.pegasys.artemis.storage.utils.UnsignedLongSerializer;
+
+public class V2MapDatabase implements Database {
+  private static final Logger LOG = LogManager.getLogger();
+  private final DB db;
+  private final Var<UnsignedLong> time;
+  private final Var<UnsignedLong> genesisTime;
+  private final Atomic.Var<Checkpoint> justifiedCheckpoint;
+  private final Atomic.Var<Checkpoint> bestJustifiedCheckpoint;
+  private final Atomic.Var<Checkpoint> finalizedCheckpoint;
+
+  private final ConcurrentNavigableMap<UnsignedLong, Bytes32> finalizedRootsBySlot;
+  private final ConcurrentMap<Bytes32, BeaconBlock> finalizedBlocksByRoot;
+  private final ConcurrentMap<Bytes32, BeaconState> finalizedStatesByRoot;
+  private final ConcurrentMap<Bytes32, BeaconBlock> hotBlocksByRoot;
+  private final ConcurrentMap<Bytes32, BeaconState> hotStatesByRoot;
+
+  private final ConcurrentMap<Checkpoint, BeaconState> checkpointStates;
+  private final ConcurrentMap<UnsignedLong, Checkpoint> latestMessages;
+
+  // In memory only
+  private final ConcurrentNavigableMap<UnsignedLong, Set<Bytes32>> hotRootsBySlotCache =
+      new ConcurrentSkipListMap<>();
+
+  public static Database createOnDisk(final File directory, final boolean startFromDisk) {
+    final File databaseFile = new File(directory, "artemis.db");
+    try {
+      if (!startFromDisk) {
+        Files.deleteIfExists(databaseFile.toPath());
```

Comment:
Seems a little scary to delete the db based on one little flag :D

Reply:
It is, there are other tickets to remove that flag - this is just inline with the existing behaviour.